### DISCUSSION TITLE
MCMC posterior distributions problem (NM fit perfect)

### DISCUSSION THREAD

**Discussion by gokhannY** [phoebe-project/phoebe2#955] [General]
I have a weird problem. My NM fit is perfect for finding appropriate parameters, but MCMC posteriors "hit a wall" as described in other discussion

This is the NM fit results
[nmfit.pdf](https://github.com/user-attachments/files/17277487/nmfit.pdf)

This is the MCMC fit results
[lcrv.pdf](https://github.com/user-attachments/files/17277493/lcrv.pdf)

Somehow mass ratio and asini posteriors go to wrong direction and stuck in there, hence hitting a wall.

Here is the failed posteriors
[failed.pdf](https://github.com/user-attachments/files/17277494/failed.pdf)

and lnprobability graph

[lnprob.pdf](https://github.com/user-attachments/files/17277497/lnprob.pdf)

What can I do to fix this?

Here is the NM fit results
```

    requiv@primary@star@component: 4.443249539070683 solRad
   teffratio@binary@orbit@comp...: 0.6957623787199289
      incl@binary@orbit@component: 70.22895828050298 deg
       pblum@primary@V@lc@dataset: 6.8733465255469035 W
         q@binary@orbit@component: 0.4671437731594412
     asini@binary@orbit@component: 15.992735820875236 solRad

```
Lastly this is the MCMC graph
[abc123.pdf](https://github.com/user-attachments/files/17277529/abc123.pdf)

Forgot to mention that this is a semi-detached system with secondary filled Roche.

**Comment by amiszuda** [comment]
In general, whether using PHOEBE or another code, the key to understanding the dataset-dependent values of  q  and  asini  lies in the data itself. While these parameters primarily manifest in radial velocities (RVs), they do influence light curves—albeit to a lesser extent. Given the significantly larger quantity of photometric data compared to RVs, as well as the respective weighting of each, it’s natural for the photometric solution to have a stronger influence on these values. I’m not certain if PHOEBE automatically handles such cases by assigning different weights to balance the impact of each dataset on the respective parameters (@kecnry @aprsa), but in my experience, fixing  q  and  asini  to their RV-determined values and adding small priors during the MCMC process does the trick.

**Comment by gokhannY** [comment]
I agree with you. That's why my Gaussian around 0.1 and 1 for q and asini, respectively. I might try 0.02 for q, but the MCM calculation takes almost five days for this system.  

I also attached the trace graph. The starting points are all over the place. The posteriors just stuck for some values for q and asini.

I might try to solve each data, rv and lc, independently in MCMC but that wouldn't be scientific.

[trace.pdf](https://github.com/user-attachments/files/17278048/trace.pdf)

**Comment by kecnry** [comment]
I agree with @amiszuda - you have a few options within phoebe to address this:
* find what other parameter(s) _might_ be able to account for the LC residuals that are "incorrectly" driving `asini` and `q`.  This may or may not be possible depending on the situation.
* account for potential underestimation of the uncertainties on the photometry points, using `sigmas_lnf` (I guess we don't have a great tutorial on this still, but it is explained in the [2.3 release paper](https://phoebe-project.org/publications/2020Conroy+)).  I'm not sure it will completely avoid the problem you're seeing, though.
* do an MCMC run on the RVs alone and then hold `q` and `asini` fixed for the MCMC run on LCs+RVs.  You can flip the constraint so that the second run can still vary the inclination while conserving `asini`.
* better yet (in my opinion), do an MCMC run on the RVs alone, and use those _posteriors_ as the _priors_ on the following MCMC run on LCs alone.  This effectively treats the information from all the RVs equally as the information from all the LCs.
* use a [custom cost function](https://phoebe-project.org/docs/latest/tutorials/emcee_custom_lnprob) to manually override the weights of each point (I'm not a fan of this one, but it is technically possible).

Whatever you end up doing, since this is somewhat "non-standard", I think its important to be very clear in any publication of exactly what you did and why so that the results are reproducible and taken in the correct context.

**Reply by kecnry** [reply]
(I also converted this from an issue to a discussion so that people could find it easier - closed issues can be easily lost)

**Reply by gokhannY** [reply]
Thanks Kyle. I used the advice of @amiszuda and started the MCMC run for asini and q with smaller distributions as 0.5 and 0.02, respectively.

I believe your 4th suggestion is the best one among them. 

https://phoebe-project.org/docs/2.4/tutorials/emcee_resampling

I believe this tutorial would satisfy my needs, if I'm not mistaken.

**Comment by gokhannY** [comment]
Firstly, sorry for delayed reply, but doing MCMC individually for RVs and LC did not work. As can be seen from the figures I attached. 

This is for RVs alone and it's good as expected. 
[RV.pdf](https://github.com/user-attachments/files/17632381/RV.pdf)

This is for LC included, as you can see _q_ and _a_ goes to different way.
[RV_LC.pdf](https://github.com/user-attachments/files/17632392/RV_LC.pdf)

And this is the trace figure for each parameter.
[RV_LC_trace.pdf](https://github.com/user-attachments/files/17632396/RV_LC_trace.pdf)

The weird thing, when I do the same analyses in Phoebe 1.0, everything is perfect. But I do not want to use Phoebe 1.0 just because it gives the expected result. 
[Phoebe1.pdf](https://github.com/user-attachments/files/17632404/Phoebe1.pdf)

I'm also attaching the code I used for secondary MCMC analyses that I included LC.

```
import phoebe
import numpy as np

b = phoebe.load('result_RV.bundle', import_from_newer=True)

b.enable_dataset('V')

b.set_value('lnprob_cutoff', solution='round_1', value=145) 

b.adopt_solution('round_1', adopt_values=False, adopt_distributions=True, distribution='resample')

b.set_value('init_from', solver='mcmc_last', value='resample')

b.add_distribution({'teffratio@binary': phoebe.gaussian_around(0.1),
    'requiv@primary': phoebe.gaussian_around(0.5),
    'incl@binary': phoebe.gaussian_around(2.0),
    'pblum@primary': phoebe.gaussian_around(2.)
    }, phoebe.gaussian_around(0.5), distribution='init_from_new')
    

b.set_value(qualifier='init_from', solver='mcmc_last', value=['resample', 'init_from_new'])

b.run_solver('mcmc_last', solution='round_2')

b.save('result_RV_LCs.bundle')
```

I'm guessing I do something wrong when I prepare the file.

**Reply by kecnry** [reply]
Sorry for the confusion - for this approach you'll want to apply the posteriors from the RVs as `priors` (not `init_from` - although using them for both would be reasonable) during the LC run.

**Comment by gokhannY** [comment]
Finally fixed, thanks for help guys. All I had to do was change this

`b.set_value('init_from', solver='mcmc_last', value='resample')`

to this

`b.set_value('priors', solver='mcmc_last', value='resample')`

[Last.pdf](https://github.com/user-attachments/files/17698180/Last.pdf)

Hope, this would help to someone else.

**Comment by gokhannY** [comment]
Sorry for re-openning this, but I'm puzzled with lately with a weird problem. The following code that I use was working until lately, when I have tried for a different system now i'm getting this error code.

```
gokhan@gkhn [16:31:18]:~$ python mcmc_son.py 
Traceback (most recent call last):
  File "/home/gokhan/mcmc_son.py", line 25, in <module>
    b.adopt_solution('round_1', adopt_values=False, adopt_distributions=True, distribution='resample')
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/frontend/bundle.py", line 13891, in adopt_solution
    ps = self.add_distribution(uniqueid=uniqueid_orig, value=dist.slice(i), distribution=distribution, auto_add_figure=kwargs.get('auto_add_figure', None), check_label=distribution is not None)
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/parameters/parameters.py", line 424, in _send_if_client
    return fctn(self, *args, **kwargs)
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/frontend/bundle.py", line 8143, in add_distribution
    ret_ps += self._add_single_distribution(overwrite_individual=kwargs.get('overwrite_individual', False), return_changes=return_changes, **dist_dict)
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/frontend/bundle.py", line 7936, in _add_single_distribution
    value=value_.deepcopy(),
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/dependencies/distl/distl.py", line 600, in deepcopy
    return self.__deepcopy__(None)
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/dependencies/distl/distl.py", line 573, in __deepcopy__
    return self.__class__(**{k:v for k,v in self.to_dict().items() if k not in ['distl', 'distl.version', 'uniqueid']})
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/dependencies/distl/distl.py", line 3457, in __init__
    multivariate = from_dict(multivariate)
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/dependencies/distl/distl.py", line 146, in from_dict
    dist = getattr(_sys.modules[__name__], classname)(**kwargs)
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/dependencies/distl/distl.py", line 7551, in __init__
    super(MVSamples, self).__init__(units, labels, labels_latex, wrap_ats,
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/dependencies/distl/distl.py", line 2855, in __init__
    super(BaseMultivariateDistribution, self).__init__(*args, **kwargs)
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/dependencies/distl/distl.py", line 680, in __init__
    super(BaseDistribution, self).__init__(uniqueid=uniqueid, **kwargs)
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/dependencies/distl/distl.py", line 552, in __init__
    setattr(self, k, v)
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/dependencies/distl/distl.py", line 7565, in samples
    self._samples = is_nd_array(value)
  File "/home/gokhan/.local/lib/python3.10/site-packages/phoebe/dependencies/distl/distl.py", line 516, in is_nd_array
    raise TypeError("must be an nd array, shape={}".format(value.shape))
TypeError: must be an nd array, shape=(0,)

```

I updated phoebe to the latest version. Did something change with the latest version?

I'm adding the code and the file that I'm using so you guys can reproduce the error.
[files.zip](https://github.com/user-attachments/files/19615612/files.zip)

**Reply by kecnry** [reply]
what is the shape of the samples parameter in the solution?  Any chance it is all empty of full of nans?

**Reply by gokhannY** [reply]
Hi Kyle,
There are no Nan values in sample to my knowledge. 
As I said on m previous message/post, this is the same code in this page, which worked flawlessly before. That's why I'm puzzled. Only changing parameters are some limits around gaussian in here.

```
b.add_distribution({'teffratio@binary': phoebe.gaussian_around(0.1),
    'requiv@primary': phoebe.gaussian_around(0.5),
    'incl@binary': phoebe.gaussian_around(2.0),
    'pblum@primary': phoebe.gaussian_around(2.)
    }, phoebe.gaussian_around(0.5), distribution='init_from_new')
```

**Reply by kecnry** [reply]
Can you show the samples?  `print(b.get_value(qualifier='samples', solution='round_1'))`

**Reply by kecnry** [reply]
Or actually better yet, `phoebe.helpers.process_mcmc_chains_from_solution(b, 'round_1')` to apply the default burnin/thinning/etc options.

**Reply by gokhannY** [reply]
```
print(b.get_value(qualifier='samples', solution='round_1'))
[[[17.0049242   0.48934716 -9.23178579]
  [16.58032445  0.48162392 -6.458125  ]
  [16.22929149  0.4687254  -3.12961968]
  ...
  [16.67888146  0.48219367 -6.70675647]
  [16.50401862  0.47779855 -1.53753297]
  [16.98201348  0.46782038 -5.26202619]]

 [[17.0049242   0.48934716 -9.23178579]
  [16.58032445  0.48162392 -6.458125  ]
  [16.22882093  0.47063021 -4.15005581]
  ...
  [16.67715601  0.48148209 -6.63763188]
  [16.50401862  0.47779855 -1.53753297]
  [16.92735015  0.47041229 -5.52255216]]

 [[16.89772099  0.48479647 -7.83783143]
  [16.58032445  0.48162392 -6.458125  ]
  [16.22882093  0.47063021 -4.15005581]
  ...
  [16.5666774   0.48056407 -5.82541314]
  [16.36395367  0.47060412 -3.321027  ]
  [16.92735015  0.47041229 -5.52255216]]

 ...

 [[16.19142233  0.50054457 -3.34397906]
  [16.20538904  0.4986968  -5.71695501]
  [16.18144758  0.5012149  -7.70348579]
  ...
  [16.17954412  0.50168631 -4.73529686]
  [16.18400279  0.50092342 -5.83262535]
  [16.22601873  0.49570522 -7.83607353]]

 [[16.19142233  0.50054457 -3.34397906]
  [16.22228149  0.49679011 -5.49190852]
  [16.18830896  0.50049147 -7.48996713]
  ...
  [16.18054238  0.50158148 -4.64287737]
  [16.17382822  0.50238867 -6.01673595]
  [16.21496714  0.49721564 -7.73197272]]

 [[16.19142233  0.50054457 -3.34397906]
  [16.22146374  0.49663781 -6.27474241]
  [16.19073703  0.50016553 -7.9419097 ]
  ...
  [16.18084035  0.5015836  -4.50150995]
  [16.17317774  0.50247952 -6.05066069]
  [16.20018031  0.49928058 -8.92371771]]]
```


```
phoebe.helpers.process_mcmc_chains_from_solution(b, 'round_1')
(array([45.88035672, 47.7697722 , 47.75152481, ..., 49.04141191,
       49.90524791, 47.19823995]), array([[16.24407285,  0.49395713, -3.16877856],
       [16.19403166,  0.50082858, -4.26146797],
       [16.23664072,  0.49460068, -4.21153323],
       ...,
       [16.18084035,  0.5015836 , -4.50150995],
       [16.17317774,  0.50247952, -6.05066069],
       [16.20018031,  0.49928058, -8.92371771]]))

```

**Reply by kecnry** [reply]
I'm not really sure what's going on here, but trying to simplify the case and see which parameter/option is causing the failure would help.  Can you try removing a single parameter at a time from `adopt_parameters` and see if that ever succeeds?  And if not, try changing the value of `distributions_convert` (in the solution, before calling `adopt_solution`) and see if that works - perhaps one of the non-mutivariate options will work.

**Reply by gokhannY** [reply]
I believe I found the line that causes problem.

```
b.set_value('lnprob_cutoff', solution='round_1', value=145)
```

when I commented off this, it works. I tried to change to value bigger or lower but result was the same. I believe this parameter is a must since it eliminates to walkers that yielding to different results. Or can I just delete that line?

**Reply by kecnry** [reply]
can you still plot the lnprobability after applying the cutoff?  Do all chains still exist?

**Reply by gokhannY** [reply]
No, it doesn't show up.

Before cutoff
![before_cutoff](https://github.com/user-attachments/assets/86af86d3-a275-4254-a6ec-eec54a8f732d)

After cutoff
![after_cutoff](https://github.com/user-attachments/assets/4db3b46d-8b92-4eda-a0d1-e9e6057acfb7)

**Reply by kecnry** [reply]
ok, based on the lnprobability plot, you do not want `lnprob_cutoff` to be any value above ~35.  It looks like the applied burnin is doing its just though and so you probably can safely skip setting `lnprob_cutoff`.
